include/swift/SIL/SILDifferentiabilityWitness.h

```diff
@@ -43,78 +43,89 @@ class SILDifferentiabilityWitness
 {
 private:
   /// The module which contains the differentiability witness.
-  SILModule &module;
+  SILModule &Module;
   /// The linkage of the differentiability witness.
-  SILLinkage linkage;
+  SILLinkage Linkage;
   /// The original function.
-  SILFunction *originalFunction;
+  SILFunction *OriginalFunction;
   /// The autodiff configuration: parameter indices, result indices, derivative
   /// generic signature (optional).
-  AutoDiffConfig config;
+  AutoDiffConfig Config;
   /// The JVP (Jacobian-vector products) derivative function.
-  SILFunction *jvp;
+  SILFunction *JVP;
   /// The VJP (vector-Jacobian products) derivative function.
-  SILFunction *vjp;
+  SILFunction *VJP;
+  /// Whether or not this differentiability witness is a declaration.
+  bool IsDeclaration;
```

Comment:
I don’t feel this should be a field — when JVP and VJP and null, this witness is naturally a declaration.

Reply:
Quick comparison:
1. Make `isDeclaration` a computed property: true if JVP and VJP are null. `isDeclaration` becomes implicit and automatic, there's no way to set it explicitly. This is arguably a bit confusing for differentiability witnesses that initially have no JVP/VJP (SILGen for `@differentiable` attribute with no specified derivatives) but later have JVP and VJP set by the differentiation transform - the result of `isDeclaration` depends on whether the differentiation transform has canonicalized the witness. `SILFunction` does this (is declaration if there are no basic blocks).
2. Make `isDeclaration` a stored property. `isDeclaration` is very explicit and must be set (and maybe updated) intentionally. `SILWitnessTable` does this.

I think 1 may be preferable if possible, but I'm not sure.

Reply:
Actually, you are right about the phase-sensitivity issue.  It'll be wrong if the differentiation transform generates a witness declaration and then tries to populate it in another iteration.